mods/carts/init.lua

```diff
@@ -0,0 +1,355 @@
+		if math.abs(vel.x + vel.z) > carts.punch_speed_min then
+			return
+		end
+	end
+
+	local punch_dir = carts:velocity_to_dir(puncher:get_look_dir())
+	punch_dir.y = 0
+	local cart_dir = carts:get_rail_direction(pos, punch_dir, nil, nil, self.railtype)
+	if vector.equals(cart_dir, {x=0, y=0, z=0}) then
+		return
+	end
+
+	local punch_interval = 1
+	if tool_capabilities and tool_capabilities.full_punch_interval then
+		punch_interval = tool_capabilities.full_punch_interval
+	end
+	time_from_last_punch = math.min(time_from_last_punch or punch_interval, punch_interval)
+	local f = 3 * (time_from_last_punch / punch_interval)
+
+	self.velocity = vector.multiply(cart_dir, f)
+	self.old_dir = cart_dir
+	self.old_pos = nil
+	self.punched = true
+end
+
+function cart_entity:on_step(dtime)
+	local vel = self.object:getvelocity()
+	local update = {}
+	if self.punched then
+		vel = vector.add(vel, self.velocity)
+		self.object:setvelocity(vel)
+		self.old_dir.y = 0
+	elseif vector.equals(vel, {x=0, y=0, z=0}) then
+		return
+	end
+
+	-- dir:         New moving direction of the cart
+	-- last_switch: Currently pressed L/R key, used to ignore the key on the next rail node
+	local dir, last_switch
+	local pos = self.object:getpos()
+
+	if self.old_pos and not self.punched then
+		local flo_pos = vector.round(pos)
+		local flo_old = vector.round(self.old_pos)
+		if vector.equals(flo_pos, flo_old) then
+			-- Do not check one node multiple times
+			return
+		end
+	end
+
+	local ctrl, player
+
+	-- Get player controls
+	if self.driver then
+		player = minetest.get_player_by_name(self.driver)
+		if player then
+			ctrl = player:get_player_control()
+		end
+	end
+
+	if self.old_pos then
+		-- Detection for "skipping" nodes
+		local expected_pos = vector.add(self.old_pos, self.old_dir)
+		local found_path = carts:pathfinder(
+			pos, expected_pos, self.old_dir, ctrl, self.old_switch, self.railtype
+		)
+
+		if not found_path then
+			-- No rail found: reset back to the expected position
+			pos = expected_pos
+			update.pos = true
+		end
+	end
+
+	local cart_dir = carts:velocity_to_dir(vel)
+	local max_vel = carts.speed_max
+	dir, last_switch = carts:get_rail_direction(
+		pos, cart_dir, ctrl, self.old_switch, self.railtype)
+
+	local new_acc = {x=0, y=0, z=0}
+	if vector.equals(dir, {x=0, y=0, z=0}) then
+		vel = {x=0, y=0, z=0}
+		pos = vector.round(pos)
+		update.pos = true
+		update.vel = true
+	else
+		-- If the direction changed
+		if dir.x ~= 0 and self.old_dir.z ~= 0 then
+			vel.x = dir.x * math.abs(vel.z)
+			vel.z = 0
+			pos.z = math.floor(pos.z + 0.5)
+			update.pos = true
+		end
+		if dir.z ~= 0 and self.old_dir.x ~= 0 then
+			vel.z = dir.z * math.abs(vel.x)
+			vel.x = 0
+			pos.x = math.floor(pos.x + 0.5)
+			update.pos = true
+		end
+		-- Up, down?
+		if dir.y ~= self.old_dir.y then
+			vel.y = dir.y * math.abs(vel.x + vel.z)
+			pos = vector.round(pos)
+			update.pos = true
+		end
+
+		-- Slow down or speed up..
+		local acc = dir.y * -1.8
```

Comment:
I was considering acceleration on a slope, at 45 deg this will be 0.707 * g, 
`new_acc = vector.multiply(dir, acc)` It seems dir is not a normalised vector but can have a horizontal component 1 and a vertical componenet 1, so on a slope acc is multiplied by 1.414.
So acceleration should be set to 0.707 * 0.707 * g = 0.5 * g = 4.9.
But this is ignoring wind resistence and friction, and is almost equal to power rail acc, so how about
`dir.y * -4` as mentioned above?